marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+        demerkleizes (Input _) = True
+```
+
+
+#### *Constraint 13.* Positive balances
+
+All accounts in the state have positive balances.
+```haskell
+all ((> 0) . snd) (accounts $ marloweState marloweData) ≡ True
+```
+
+
+#### *Constraint 14.* Inputs authorized
+
+The required signature or at-least-one role token is present in the transaction for each deposit or choice.
+```haskell
+all authorized (fmap getInputContent transactionInput) ≡ True
+  where getInputContent (Input inputContent) = inputContent
+        getInputContent (MerkleizedTxInput inputContent _) = inputContent
+        authorized (IDeposit _ (PK hash) _ _) = authorizedHash hash
+        authorized (IDeposit _ (Role role) _) = authorizedRole role
+        authorized (IChoice (ChoiceId _ (PK hash)) _) = authorizedHash hash
+        authorized (IChoice (ChoiceId _ (Role role)) _) = authorizedRole role
+        authorized INotify = True
+        authorizedHash hash = elem hash (txInfoSignatories $ scriptContextTxInfo scriptContext)
+        authorizedRole role = Value.leq (Value.singleton (rolesCurrency $ marloweparams marloweData) role 1) spent
+        spent = foldMap (txOutValue . txInInfoResolved) (txInfoInputs $ scriptContextTxInfo scriptContext)
+        -- FIXME: We should write out `Value.leq` ourselves, so the specification does not depend upon Plutus's convenience functions.
+```
+
+
+#### *Constraint 15.* Sufficient payment
+
+Sufficient funds are paid to the public keys or roles.
+```haskell
+all sufficient (txOutPayments transactionOutput)
+  where sufficient (Payment _ (Party (PK hash)) value) = Value.leq value (foldMap (paidToHash hash) (txInfoOutputs $ scriptContextTxInfo scriptContext))
+                   (Payment _ (Party (Role role)) value) = any (paysToRole role value) (txInfoOutputs $ scriptContextTxInfo scriptContext)
+        -- FIXME: Should `paysToRole` test the value in *any* UTxO or the *sum* of values?
+        paysToRole (TxOut txOutAddress txOutValue (Just hash)) role value = txOutAddress == (Ledger.scriptHashAddress . rolePayoutValidatorHash $ marloweParams marloweData)
+                                                                              && lookup hash (txInfoData $ scriptContextTxInfo scriptContext) == Just (Datum $ toBuiltinData role)
+                                                                              && Value.leq value txOutValue
+        paysToRole _ _ = False
+        paidToHash hash (TxOut (Address (PubKeyCredential hash') _) txOutValue _) = if hash == hash' then txOutValue else mempty
+        -- FIXME: We should write out `Value.leq` ourselves, so the specification does not depend upon Plutus's convenience functions.
+```
+
+
+## Plutus Validator for Marlowe Payouts
+
+
+The Marlowe payout validator for roles is a Plutus script parameterized by currency symbol.
+```haskell
+rolePayoutValidator :: CurrencySymbol -> TokenName -> () -> ScriptContext -> Bool
+```
```

Comment:
Why the argument of type `()`?

Reply:
The signature of the function requires a redeemer, but we don't need one, so we just use unit.